### DISCUSSION TITLE
WMS layer other params than layers,srs

### DISCUSSION THREAD

**Discussion by radoslavirha** [visgl/deck.gl#7736] [Q&A]
Hi,

just to clarify things around new WMSLayer.

I can pass `layers`, `srs` params to WMSLayer. What about another params like `format`, `transparent` ,.. 

Another thing I see - it always creates requests with `1.1.1` version. Example WMS service I use has also `1.3.0` and I cannot use it.

Loaders can accept those params, so it's maybe where WMSLayer can be improved?

Before this layer I had all the info about WMS layer and only constructed URL (modified TileLayer with getTileData and deck gave me only bbox data).

Maybe I can help you, but I don't know if you're willing to support any improvements in WMSLayer, or if you have some API prepared for another WMS params.

Thanks

**Comment by ibgreen** [comment]
@Radik24 Thanks for the feedback. Yes the `WMSLayer` was marked experimental for this reason, we knew we still had more thinking to do about how to handle versions, additional parameters etc.

As you seem to have noticed, there is actually a somewhat ambitious effort in loaders.gl to implement support for all the major OGC Web Standards including WMS, WMTS, WFS, etc). 

One idea is that you would be able to create a loaders.gl `WMSService` class, customize it to your heart's content, and pass it as a parameter to the `WMSLayer`.

This is actually supported today (although not documented), however the `WMSService` class doesn't yet support overriding those parameters, so it may not help you, though perhaps you could subclass it and get it to do what you need by overriding methods.

Given that these OGC standards are plagued by slightly incompatible versions and implementations, the hope was always that the community would engage and help us flush out compatibility issues.

If you are seriously interested in contributing in this area, happy to engage deeper.

**Comment by brandonjpierce** [comment]
@Radik24 I was able to get around that, for now, by providing a url template for the `data` prop:

```
const template = {
  LAYERS: '{layers}',
  BBOX: '{east},{north},{west},{south}',
  X_ATTR: 'DecLon',
  Y_ATTR: 'DecLat',
  STYLES: 'cb_raster',
  POINTSIZES: '5,3,3',
  POINTSHAPES: pointShapes,
  POINTCOLORS: pointColors,
  CB_ATTR: 'MsgType',
  CB_VALS: pointValues,
  TRANSPARENT: true,
  ORDER_LAYERS: true,
  ORDER_CLASSES: true,
  FORMAT: 'image/png',
  REQUEST: 'GetMap',
  SERVICE: 'WMS',
  SRS: 'WEB_MERCATOR',
  WIDTH: '{width}',
  HEIGHT: '{height}',
  VERSION: '1.3.0',
};

const qs = reduce(template, (prev, curr, key) => {
  prev.push(`${key}=${curr}`);
  return prev;
}, []).join('&');


new WMSLayer({
  data: `${uri}/wms?${qs}`,
  serviceType: 'template',
  ...
})
```

Note that you unfortunately cannot do:

```
const qs = new URLSearchParams(template).toString()
```

As it will escape the `{}` characters.
